config/config-api/src/com/thoughtworks/go/config/BasicCruiseConfig.java

```diff
@@ -1308,26 +1309,39 @@ public PipelineConfig findPipelineUsingThisPipelineAsADependency(String pipeline
     }
 
     @Override
-    public Map<CaseInsensitiveString, List<CaseInsensitiveString>> templatesWithPipelinesForUser(String username, List<Role> roles) {
-        HashMap<CaseInsensitiveString, List<CaseInsensitiveString>> templateToPipelines = new HashMap<>();
-        for (PipelineTemplateConfig template : getTemplates()) {
-            if (isAuthorizedToAccessTemplate(new CaseInsensitiveString(username), roles, template)) {
-                templateToPipelines.put(template.name(), new ArrayList<>());
+    public List<TemplatesToPipelines> templatesWithPipelinesForUser(String username, List<Role> roles) {
```

Comment:
This entire method has 4 levels of nested iterations. Moreover, this is called once for every user. consider moving it elsewhere and may be caching some of this.